ipa-multipoint/ipa_multipoint_jni/src/lib.rs

```diff
@@ -45,130 +39,92 @@ pub extern "system" fn Java_org_hyperledger_besu_nativelib_ipamultipoint_LibIpaM
+    // But its given in big endian format.
+    // The tree_index is the last 32 bytes of the input,
+    // so we use this method to reverse its endian
+    fn reverse_last_32_bytes(arr: &mut [u8; 64]) {
+        let last_32 = &mut arr[32..];
+        last_32.reverse();
+    }
+    reverse_last_32_bytes(&mut input);
 
-    hash64(hash_input)
+    let hash = ffi_interface::get_tree_key_hash_flat_input(committer, input);
+    env.byte_array_from_slice(&hash).unwrap()
 }
 
 /// Commit receives a list of 32 byte scalars and returns a 32 byte scalar
 /// Scalar is actually the map_to_field(commitment) because we want to reuse the commitment in parent node.
 /// This is ported from rust-verkle.
 #[no_mangle]
-pub extern "system" fn Java_org_hyperledger_besu_nativelib_ipamultipoint_LibIpaMultipoint_commit(env: JNIEnv,
-                                                                                                 _class: JClass<'_>,
-                                                                                                 input: jbyteArray)
-                                                                                                 -> jbyteArray {
-    // Input should be a multiple of 32-be-bytes.
-    let inp = env.convert_byte_array(input).expect("Cannot convert jbyteArray to rust array");
-    let len = inp.len();
-    if len % 32 != 0 {
-        env.throw_new("java/lang/IllegalArgumentException", "Invalid input length. Should be a multiple of 32-bytes.")
-           .expect("Failed to throw exception");
-        return std::ptr::null_mut(); // Return null pointer to indicate an error
-    }    
-    let n_scalars = len / 32;
-    if n_scalars > 256 {
-        env.throw_new("java/lang/IllegalArgumentException", "Invalid input length. Should be at most 256 elements of 32-bytes.")
-           .expect("Failed to throw exception");
-        return std::ptr::null_mut(); // Return null pointer to indicate an error
-    }    
-
-    // Each 32-be-bytes are interpreted as field elements.
-    let mut scalars: Vec<Fr> = Vec::with_capacity(n_scalars);
-    for b in inp.chunks(32) {
-        scalars.push(Fr::from_be_bytes_mod_order(b));
-    }
-    
-    // Committing all values at once.
-    let bases = CRS::new(n_scalars, PEDERSEN_SEED);
-    let commit = multi_scalar_mul(&bases.G, &scalars);
-
-    // Serializing via x/y in projective coordinates, to int and to scalars.
-    let scalar = group_to_field(&commit);
-    let mut scalar_bytes = [0u8; 32];
-    scalar.serialize(&mut scalar_bytes[..]).expect("could not serialise Fr into a 32 byte array");
-    scalar_bytes.reverse();
-
-    return env.byte_array_from_slice(&scalar_bytes).expect("Couldn't convert to byte array");
-}
+pub extern "system" fn Java_org_hyperledger_besu_nativelib_ipamultipoint_LibIpaMultipoint_commit(
+    env: JNIEnv,
+    _class: JClass<'_>,
+    input: jbyteArray,
+) -> jbyteArray {
+    let mut input = env
```

Comment:
Should we use the same way to read inputs as line 46 ? It is safer I believe
let mut input: [u8; 64] = match input.try_into() {
        Ok(input) => input,
        Err(_) => {
            env.throw_new(
                "java/lang/IllegalArgumentException",
                "Invalid input length. Should be 64-bytes.",
            )
            .expect("Failed to throw exception");
            return std::ptr::null_mut(); // Return null pointer to indicate an error
        }
    };

Reply:
Ah, I missed that!
Nonetheless, it is strange that in one case we throw a Java Error and in the other we don't.
However, I am not sure how error propagation works on the JNI side.
I think that we can do as is and if the need emerge to handle errors in another way, we will do it then.

Reply:
I will prefer to  use 
```
env.throw_new(
                "java/lang/IllegalArgumentException",
                "message",
            )
            .expect("Failed to throw exception");
```
 everytime in order to be sure java will catch this cleanly imo

Reply:
as it's not related to this PR I think we can merge it